data/lua/cave_map_generator.lua

```diff
@@ -1,200 +1,184 @@
+			if item.place_castle then
+				map:set_tile(x, y, params.terrain_keep)
+				for x2, y2 in MG.adjacent_tiles(x, y) do
+					map:set_tile(x2, y2, params.terrain_castle)
+				end
 			end
 		end
 	end
-end
 
-for i,v in ipairs(passages) do
-	if random(100) <= (v.data.chance or 100) then
-		local windiness = v.data.windiness or 0
-		local laziness = math.max(v.data.laziness or 1, 1)
-		local width = math.max(v.data.width or 1, 1)
-		local jagged = v.data.jagged or 0
-		local calc = function(x, y, current_cost)
-			local res = 1.0
-			if get_tile(x, y) == params.terrain_wall then
-				res = laziness
+	for i,v in ipairs(passages) do
+		if random(100) <= (v.data.chance or 100) then
+			local windiness = v.data.windiness or 0
+			local laziness = math.max(v.data.laziness or 1, 1)
+			local width = math.max(v.data.width or 1, 1)
+			local jagged = v.data.jagged or 0
+			local calc = function(x, y, current_cost)
+				local res = 1.0
+				if map:get_tile(x, y) == params.terrain_wall then
+					res = laziness
+				end
+				if windiness > 1 then
+					res = res * random(windiness)
+				end
+				return res
 			end
-			if windiness > 1 then
-				res = res * random(windiness)
+			local path, cost = wesnoth.find_path(v.start_x, v.start_y, v.dest_x, v.dest_y, calc, params.map_width, params.map_height)
+			for i, loc in ipairs(path) do
+				local locs_set = LS.create()
+				build_chamber(loc[1], loc[2], locs_set, width, jagged)
+				for k,v in ipairs(locs_set:to_pairs()) do
+					clear_tile(v[1], v[2])
+				end
 			end
-			return res
 		end
-		local path, cost = wesnoth.find_path(v.start_x, v.start_y, v.dest_x, v.dest_y, calc, params.map_width, params.map_height)
-		for i, loc in ipairs(path) do
-			local locs_set = {}
-			build_chamber(loc[1], loc[2], locs_set, width, jagged)
-			for k,v in pairs(locs_set) do
-				set_tile_index(k, params.terrain_clear)
+
+	end
+
+	if type(params.transform) == "string" then
+		local chance = params.transform_chance or 100
+		if random(100) <= chance then
+			local transforms = {}
+			for t in params.transform:gmatch("[^%s,][^,]*") do
+				-- TODO: Possibly support other transformations, such as transpose, rotate_cw, rotate_ccw, etc
+				if t == 'flip_x' or t == 'flip_y' then
+					table.insert(transforms, t)
+				else
+					error("Unknown transformation '" .. t .. "'")
+				end
 			end
+			map[transforms[random(#transforms)]](map)
 		end
 	end
-	
-end
 
-local stringbuilder = {}
+	return tostring(map)
+end
 
-for y = 0, params.map_height - 1 do
-	local stringbuilder_row = {}
-	for x = 0, params.map_width - 1 do
-		local index = x + 1 + y * params.map_width
-		if starting_positions[index] ~= nil then
-			table.insert(stringbuilder_row, tostring(starting_positions[index]) .. " " .. map[index])
-			else
-			table.insert(stringbuilder_row, map[index])
+function callbacks.generate_scenario(params)
+	-- This is more or less backwards compatible with the cave generator syntax
+	local scenario = helper.child(params, "settings")
```

Comment:
I think this shodul cal called `[scenario]` instead of `[settings]` mostly becasue it also called `[scenario]` in the defautl map generator